File: vs/CSharpCompiler/Lab1/Lab5.cs
using System;
using System.Collections.Generic;
using System.Drawing;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Lab1
{
    public class Lab5
    {
        public static bool checking_string_validity(string url)
        {
            return checking_string_validity(url, out string t);
        }
        public static string get_domain(string url)
        {
            string dom = "";
            if(!checking_string_validity(url,out  dom))
            {
                throw new Exception("Некорректное имя домена!");
            }
            //http
            char[] ch = url.ToLower().ToCharArray();
            int strat_pos = 4;
            if (ch[strat_pos] == 's')
                strat_pos++;
            strat_pos += 3;
            dom = "";
            while (ch[strat_pos]!='/')
            {
                dom += ch[strat_pos++].ToString();
            }
            return dom;
        }
        private static bool checking_string_validity(string url,out string domain_)
        {
            string start_val = "http";
            domain_ = "";
            if(url==null||url.Length<=start_val.Length)
                return false;
            string tmp = "";
            string http_no_secure = "http:";
            string http_secure = "https:";
            string domain = "";
            int last_point_position = 0;
            int pos = 0;
            int flag1 = 0,flag2=0,flag3=0;
            string domain_zone = "";
            string the_rest_of_the_way = "";
            foreach (char c in url.ToLower().ToCharArray())
            {
                if (pos < start_val.Length)
                {
                      tmp += c.ToString();
                      pos++;
                }
                else if(c!='/' && flag1==0)
                {
                    {
                        if (c != 's' && c != ':')
                            return false;
                        else
                        {
                            tmp += c.ToString();
                            pos++;
                            if (c == ':')
                            {
                                if (tmp != http_no_secure && tmp != http_secure)
                                    return false;
                            }
                        }
                    }
                }
                else
                {
                    tmp += c.ToString();
                    pos++;
                    if (c == '/' && flag1==0)
                    {
                        flag1 = 1;
                        if (tmp != (http_no_secure + '/'.ToString()) && tmp != (http_secure + "/".ToString()))
                        {
                            return false;
                        }
                    }
                    else if (flag1 == 1 && flag2 == 0)
                    {
                        last_point_position = -1;
                        int pos2 = pos+1;
                        char[]c_url=url.ToCharArray();
                        while ((pos2 < c_url.Length) && (c_url[pos2]!='/'))
                        {
                            if (c_url[pos2] == '.')
                            {
                                last_point_position = pos2;
                            }
                            pos2++;
                        }
                        if (last_point_position < 0)
                            return false;
                        flag2 = 2;
                    }
                    else 
                    {
                        if ((pos < last_point_position)&& c!='/')
                        {
                             domain += c.ToString();
                        }
                        else if(pos==last_point_position)
                        {
                             string table = "qwertyuiopasdfghjklzxcvbnm.1234567890-";
                             foreach (char v in domain.ToCharArray())
                             if (table.Contains(v) == false)
                                 return false;
                        }
                        else 
                        {
                            if (flag3 == 0)
                            {
                                if (c != '/')
                                {
                                    if (c != '.')
                                        domain_zone += c.ToString();
                                }
                                else
                                {
                                    string table_domain = "qwertyuiopasdfghjklzxcvbnm";
                                    foreach (char v in domain_zone.ToCharArray())
                                        if (table_domain.Contains(v) == false)
                                            return false;
                                    flag3 = 1;
                                }
                            }
                            else
                            {
                                the_rest_of_the_way += c.ToString();
                            }
                        }
                        
                    }
                    
                    
                }
            }
            if (the_rest_of_the_way.Length > 0)
            {
                string url_full_char_table = "qwertyuiopasdfghjklzxcvbnm/?&+;#1234567890-=._[]";
                foreach (char c in the_rest_of_the_way.ToCharArray())
                    if (url_full_char_table.Contains(c) == false)
                        return false;
            }
            domain_ = domain+"."  + domain_zone;
            return true;
        }
    }
}

File: vs/PythonApplications/Lab5.py
'''
Реализовать функцию, которая будет проверять, является ли 
введенная строка доменом из URL-адреса, возвращаемое значение True или 
False. Дополнительно реализовать функцию, которая выбрасывает исключение 
о некорректном аргументе, иначе возвращает домен из URL-адреса.
Доменные имена для протоколов http и https, с необязательным слешем 
в конце. Специальные символы не используются.
Примеры:
http://example.com/ – да
example.com – нет
кремль.рф – нет :(
'''

import re

def checking_string_validity_(url, domain_=None):
    start_val = "http"
    domain_ = ""
    if url is None or len(url) <= len(start_val):
        return (False,domain_)
    tmp = ""
    http_no_secure = "http:"
    http_secure = "https:"
    domain = ""
    last_point_position = 0
    pos = 0
    flag1 = 0
    flag2 = 0
    flag3 = 0
    domain_zone = ""
    the_rest_of_the_way = ""
    for c in url.lower():
        if pos < len(start_val):
            tmp += c
            pos += 1
        elif c != '/' and flag1 == 0:
            if c != 's' and c != ':':
                return (False,domain_)
            else:
                tmp += c
                pos += 1
                if c == ':':
                    if tmp != http_no_secure and tmp != http_secure:
                        return (False,domain_)
        else:
            tmp += c
            pos += 1
            if c == '/' and flag1 == 0:
                flag1 = 1
                if tmp != (http_no_secure + '/') and tmp != (http_secure + "/"):
                    return (False,domain_)
            elif flag1 == 1 and flag2 == 0:
                last_point_position = -1
                pos2 = pos
                while pos2 < len(url) and url[pos2] != '/':
                    if url[pos2] == '.':
                        last_point_position = pos2
                    pos2 += 1
                if last_point_position < 0:
                    return (False,domain_)
                flag2 = 2
            else:
                if pos < last_point_position and c != '/':
                    domain += c
                elif pos == last_point_position:
                    table = "qwertyuiopasdfghjklzxcvbnm.1234567890-"
                    for v in domain:
                        if v not in table:
                            return (False,domain_)
                else:
                    if flag3 == 0:
                        if c != '/':
                            if c != '.':
                                domain_zone += c
                        else:
                            table_domain = "qwertyuiopasdfghjklzxcvbnm"
                            for v in domain_zone:
                                if v not in table_domain:
                                    return (False,domain_)
                            flag3 = 1
                    else:
                        the_rest_of_the_way += c
    if the_rest_of_the_way:
        url_full_char_table = "qwertyuiopasdfghjklzxcvbnm/?&+;#1234567890-=._[]"
        for c in the_rest_of_the_way:
            if c not in url_full_char_table:
                return (False,domain_)
    domain_ = domain + "." + domain_zone
    return (True,domain_)

def checking_string_validity(url):
    (res,domain_)= checking_string_validity_(url,domain_=None)
    return res

def get_domain(url):
    
        (res,_domain_)= checking_string_validity_(url,domain_=None)
        if(res!=True):
            raise Exception("Некорректное имя домена!")
        

        url = url.lower()
        start_pos = 4
        if url[start_pos] == 's':
           start_pos += 1
        start_pos += 3
        domain = ""
        while url[start_pos] != '/':
              domain += url[start_pos]
              start_pos += 1
        _domain_=domain

        return _domain_
   


urls = [
    "https://www.google.com/search?q=base64+code+table&oq=base64+code+ta&gs_lcrp=EgZjaHJvbWUqBwgBEAAYgAQyBggAEEUYOTIHCAEQABiABDIKCAIQABiABBiiBDIKCAMQABiABBiiBDIKCAQQABiABBiiBDIKCAUQABiABBiiBDIKCAYQABiABBiiBNIBCjIyNjU1ajBqMTWoAgiwAgE&sourceid=chrome&ie=UTF-8",
    "http://example.com/",
    "example.com",
    "погода.ру",
    "http://qwert.@@@@/g.ru",
    "https://www.dns-shop.ru/",
    "https://www.dns-shop.ru/catalog/17a89a0416404e77/materinskie-platy/?stock=now-today-tomorrow-later&f[rv2z]=13iyb1&f[uq]=lxihi",
    "https://dotnet.microsoft.comen-us/apps/aspnet",
    "https://learn.microsoft.com/en-us/aspnet/core/blazor/security/?view=aspnetcore-8.0&WT.mc_id=dotnet-35129-website"
]

for url in urls:
    print(url)
    print("\n")
    print(checking_string_validity(url))
    try:
        print(get_domain(url))
    except Exception as e:
        print("error:")
        print(e)

